lib/services/cloud-service.ts

```diff
@@ -1,89 +1,108 @@
-			return new Promise<IServerStatus>(async (resolve, reject) => {
-				try {
-					resolve(await this.getObjectFromS3File<IServerStatus>(serverInformation.statusUrl));
-				} catch (err) {
-					this.$logger.trace(err);
-					reject(new Error(this.failedToStartError));
+	protected async waitForServerOperationToFinish(cloudOperationId: string, serverResponse: IServerResponse, options: ICloudOperationExecutionOptions): Promise<ICloudOperationResult> {
+		const cloudOperationVersion = serverResponse.cloudOperationVersion || CloudService.CLOUD_OPERATION_VERSION_1;
+		const cloudOperation: ICloudOperation = this.$nsCloudOperationFactory.create(cloudOperationVersion, cloudOperationId, serverResponse);
+		this.cloudOperations[cloudOperationId] = cloudOperation;
+
+		cloudOperation.on(CloudCommunicationEvents.MESSAGE, (m: ICloudOperationMessage<any>) => {
+			if (m.type === CloudOperationMessageTypes.CLOUD_OPERATION_OUTPUT && !options.silent) {
+				const body: ICloudOperationOutput = m.body;
+				let log = body.data;
+				if (cloudOperationVersion !== CloudService.CLOUD_OPERATION_VERSION_1) {
+					log = this.$nsCloudOutputFilter.filterBpcMetadata(log);
 				}
 
-			}).catch(retry);
-		}, promiseRetryOptions)
-			.then((serverStatus: IServerStatus) => {
-				return new Promise<void>((resolve, reject) => {
-					this.outputCursorPosition = 0;
-					const serverIntervalId = setInterval(async () => {
-						if (serverStatus.status === CloudService.OPERATION_COMPLETE_STATUS) {
-							await this.getServerLogs(serverInformation.outputUrl, operationId);
-							clearInterval(serverIntervalId);
-							return resolve();
-						}
-
-						if (serverStatus.status === CloudService.OPERATION_FAILED_STATUS) {
-							await this.getServerLogs(serverInformation.outputUrl, operationId);
-							clearInterval(serverIntervalId);
-							return reject(new Error(this.failedError));
-						}
-
-						if (serverStatus.status === CloudService.OPERATION_IN_PROGRESS_STATUS_OLD ||
-							serverStatus.status === CloudService.OPERATION_IN_PROGRESS_STATUS) {
-							await this.getServerLogs(serverInformation.outputUrl, operationId);
-						}
-
-						serverStatus = await this.getObjectFromS3File<IServerStatus>(serverInformation.statusUrl);
-					}, CloudService.OPERATION_STATUS_CHECK_INTERVAL);
-				});
-			});
+				if (body.pipe === "stdout") {
+					// Print the output on the same line to have cool effects like loading indicators.
+					// The cloud process will take care of the new lines.
+					this.$logger.printInfoMessageOnSameLine(log);
+				} else if (body.pipe === "stderr") {
+					this.$logger.error(log);
+				}
+			}
+
+			this.emit(CloudCommunicationEvents.MESSAGE, m);
+		});
+
+		try {
+			await cloudOperation.init();
+		} catch (err) {
+			this.$logger.trace(err);
+			await cloudOperation.cleanup();
+			throw new Error(this.failedToStartError);
+		}
+
+		try {
+			const result = await cloudOperation.waitForResult();
+			await cloudOperation.cleanup();
```

Comment:
When the operation completes, should we remove it from this.cloudOperations?

Reply:
Yes we should.

Reply:
Yes. I am removing it in the executeCloudOperation method now.